File: ATTT_Java/ATTT/src/attt/Vegenere.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package attt;
import java.util.Scanner;
/**
 *
 * @author Admin
 */
public class Vegenere {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Nhập vào văn bản cần mã hóa: ");
        String plainText = scanner.nextLine().toUpperCase();

        System.out.print("Nhập vào khóa: ");
        String key = scanner.nextLine().toUpperCase();

        String cipherText = encrypt(plainText, key);
        System.out.println("Văn bản đã mã hóa: " + cipherText);

        String decryptedText = decrypt(cipherText, key); // Thêm phần giải mã
        System.out.println("Văn bản đã giải mã: " + decryptedText);

        scanner.close();
    }

    public static String encrypt(String plainText, String key) {
        String cipherText = ""; // Khởi tạo chuỗi rỗng để lưu văn bản đã mã hóa
        int keyIndex = 0;

        for (int i = 0; i < plainText.length(); i++) {
            char plainChar = plainText.charAt(i);
            char keyChar = key.charAt(keyIndex);

            if (Character.isLetter(plainChar)) {
                // Tìm vị trí của ký tự trong bảng chữ cái (bằng ASCII)
                int plainCharIndex = plainChar - 'A';
                int keyCharIndex = keyChar - 'A';

                // Thực hiện mã hóa Vigenere
                int cipherCharIndex = (plainCharIndex + keyCharIndex) % 26;
                char cipherChar = (char) (cipherCharIndex + 'A');
                cipherText += cipherChar; // Nối thêm ký tự mã hóa vào chuỗi cipherText
            } else {
                cipherText += plainChar; // Giữ nguyên ký tự không phải chữ cái
            }

            keyIndex = (keyIndex + 1) % key.length(); // Di chuyển đến ký tự tiếp theo trong khóa
        }
        return cipherText; // Trả về chuỗi văn bản đã mã hóa
    }

    public static String decrypt(String cipherText, String key) { // Hàm giải mã
        String decryptedText = ""; // Khởi tạo chuỗi rỗng để lưu văn bản đã giải mã
        int keyIndex = 0;

        for (int i = 0; i < cipherText.length(); i++) {
            char cipherChar = cipherText.charAt(i);
            char keyChar = key.charAt(keyIndex);

            if (Character.isLetter(cipherChar)) {
                // Tìm vị trí của ký tự trong bảng chữ cái (bằng ASCII)
                int cipherCharIndex = cipherChar - 'A';
                int keyCharIndex = keyChar - 'A';

                // Thực hiện giải mã Vigenere
                int decryptedCharIndex = (cipherCharIndex - keyCharIndex + 26) % 26; // Trừ mã số và tính modulo
                char decryptedChar = (char) (decryptedCharIndex + 'A');
                decryptedText += decryptedChar; // Nối thêm ký tự giải mã vào chuỗi decryptedText
            } else {
                decryptedText += cipherChar; // Giữ nguyên ký tự không phải chữ cái
            }

            keyIndex = (keyIndex + 1) % key.length(); // Di chuyển đến ký tự tiếp theo trong khóa
        }

        return decryptedText; // Trả về chuỗi văn bản đã giải mã
    }
}

File: ATTT_C#/ATTT/Vegenere.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace ATTT
{
    public partial class Vegenere : Form
    {
        public Vegenere()
        {
            InitializeComponent();
        }

        private void btn_mh_Click(object sender, EventArgs e)
        {
            // Lấy văn bản và khóa, chuyển thành chữ in hoa
            string plainText = txtInput.Text.ToUpper();
            string key = txtKey.Text.ToUpper();

            // Kiểm tra khóa
            if (string.IsNullOrWhiteSpace(key))
            {
                MessageBox.Show("Vui lòng nhập khóa!", "Lỗi", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            // Mã hóa văn bản
            string cipherText = Encrypt(plainText, key);
            txtOutput.Text = cipherText;
        }

        private void btn_gm_Click(object sender, EventArgs e)
        {
            // Lấy văn bản và khóa, chuyển thành chữ in hoa
            string cipherText = txtInput.Text.ToUpper();
            string key = txtKey.Text.ToUpper();

            // Kiểm tra khóa
            if (string.IsNullOrWhiteSpace(key))
            {
                MessageBox.Show("Vui lòng nhập khóa!", "Lỗi", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            // Giải mã văn bản
            string decryptedText = Decrypt(cipherText, key);
            txtOutput.Text = decryptedText;
        }

        // Hàm mã hóa Vigenère
        public static string Encrypt(string plainText, string key)
        {
            StringBuilder cipherText = new StringBuilder();
            int keyIndex = 0;

            foreach (char plainChar in plainText)
            {
                if (char.IsLetter(plainChar))
                {
                    // Tìm vị trí của ký tự trong bảng chữ cái
                    int plainCharIndex = plainChar - 'A';
                    int keyCharIndex = key[keyIndex] - 'A';

                    // Thực hiện mã hóa Vigenère
                    int cipherCharIndex = (plainCharIndex + keyCharIndex) % 26;
                    char cipherChar = (char)(cipherCharIndex + 'A');
                    cipherText.Append(cipherChar);

                    // Di chuyển đến ký tự tiếp theo trong khóa
                    keyIndex = (keyIndex + 1) % key.Length;
                }
                else
                {
                    // Giữ nguyên ký tự không phải chữ cái
                    cipherText.Append(plainChar);
                }
            }

            return cipherText.ToString();
        }

        // Hàm giải mã Vigenère
        public static string Decrypt(string cipherText, string key)
        {
            StringBuilder decryptedText = new StringBuilder();
            int keyIndex = 0;

            foreach (char cipherChar in cipherText)
            {
                if (char.IsLetter(cipherChar))
                {
                    // Tìm vị trí của ký tự trong bảng chữ cái
                    int cipherCharIndex = cipherChar - 'A';
                    int keyCharIndex = key[keyIndex] - 'A';

                    // Thực hiện giải mã Vigenère
                    int decryptedCharIndex = (cipherCharIndex - keyCharIndex + 26) % 26;
                    char decryptedChar = (char)(decryptedCharIndex + 'A');
                    decryptedText.Append(decryptedChar);

                    // Di chuyển đến ký tự tiếp theo trong khóa
                    keyIndex = (keyIndex + 1) % key.Length;
                }
                else
                {
                    // Giữ nguyên ký tự không phải chữ cái
                    decryptedText.Append(cipherChar);
                }
            }

            return decryptedText.ToString();
        }
    }
}
